### DISCUSSION TITLE
v2.10.0

### DISCUSSION THREAD

**Discussion by Tochemey** [Tochemey/goakt#558] [Announcements]
## What's Changed

- Refactor supervisor strategy. See breaking change note.
- Modify testkit method `Send` signature to enforce that testing remains in the probe context by using the actor name instead of the PID
- New testkit method `SendSync` to assert Ask-pattern messages has been added.
- Actor restart implementation has been refactored. This is the new restart behavior:
    - Restart the given actor
    - Restart all direct children of the given actor that are alive before the actor is restarted. Grand children are excluded.
    - Respect hierarchical relationship between the direct children and the restarted actor in a parent-child relation.
    - The restarted actor should be added back to its parent if any. 
- Message processing implementation has been refactored to enhance the throughput of number of messages to handle and reduce resource consumption.
- _**Unbounded priority mailbox**_ added. Performance wise it seems ok for most use cases. Further work may enhance its performance.

## BREAKING CHANGE

The main breaking in this release is how the supervisor strategy work now. The developer can easily create mappings between error types and supervisor strategy directives via the `NewSupervisorStrategy` method and `SpawnOption`. This allows fine-grained error handling during message processing by actor. In that regard the same actor can have one more ways of dealing with error types. Supervisor strategies are defined during the creation of actors. Therefore a faulty actor can be handled differently depending upon the error type. For instance with this new enhancement one can `restart` an actor with the following error type `&runtime.PanicNilError{}` and the same actor to be stopped on a different error type. More information on the readme.

The following methods are removed:

- `WithSupervisorDirective`: removed from the actor system option
- `WithSupervisor`: removed from the SpawnOption

## Pull requests

* feat: implement priority mailbox by @Tochemey in https://github.com/Tochemey/goakt/pull/546
* refactor: refactor internal engine by @Tochemey in https://github.com/Tochemey/goakt/pull/548
* fix: fix actor restarting implementation by @Tochemey in https://github.com/Tochemey/goakt/pull/550
* perf: enhance message processing by @Tochemey in https://github.com/Tochemey/goakt/pull/551
* refactor: refactor implementation of supervisor strategy by @Tochemey in https://github.com/Tochemey/goakt/pull/553
* feat: add ask-pattern testing mechanism by @Tochemey in https://github.com/Tochemey/goakt/pull/555

**Full Changelog**: https://github.com/Tochemey/goakt/compare/v2.9.1...v2.10.0

<hr /><em>This discussion was created from the release <a href='https://github.com/Tochemey/goakt/releases/tag/v2.10.0'>v2.10.0</a>.</em>

**Comment by incizzle** [comment]
👀 I see you added `runtime.Gosched()` I saw this mentioned in https://www.youtube.com/watch?v=aKz2xTVrYXc about his actor model pkg. 

Were you able to run any benchmarks around this?

**Reply by Tochemey** [reply]
@incizzle interesting that there is another actor framework out there :). Now it seems we have a lot in the go community, developers are spoilt of choices:
- https://github.com/ergo-services/ergo
- https://github.com/anthdm/hollywood
- https://github.com/asynkron/protoactor-go (I was a contributor some time past)

**Reply by Tochemey** [reply]
However runtime.Gosched() is just to yield process back to the other go routines. https://pkg.go.dev/runtime#Gosched. Regarding benchmarking. the latest changes enhanced a bit the processing of messages. However more enhancement need to be done. I am also open to any new suggestions and PRs. cc @incizzle

**Reply by incizzle** [reply]
Very much spoilt with choice goakt has been my favourite for sure. Good to know looking forward to future performance and feature updates thanks for your hard work 🫡

**Reply by Tochemey** [reply]
@incizzle Thanks for using GoAkt. I hope it meets all your needs. I am adding a new feature called `StatefulActor` to help build CRUD with GoAkt with durable state and versioning. There is already a draft PR. I am still working on it.
